dali/test/python/test_operator_noise_salt_and_pepper.py

```diff
@@ -0,0 +1,131 @@
+    pixel_mask = np.logical_and(np.all(input != pepper_val, axis=-1),
+                                np.all(input != salt_val, axis=-1))
+    salt_count = np.count_nonzero(np.logical_and(salt_mask, pixel_mask))
+    pepper_count = np.count_nonzero(np.logical_and(pepper_mask, pixel_mask))
+    pixel_count = np.count_nonzero(pixel_mask)
+    assert (np.logical_or(passthrough_mask, np.logical_or(salt_mask, pepper_mask))).all()
+    actual_noise_prob = (pepper_count + salt_count) / pixel_count
+    actual_salt_to_pepper_prob = salt_count / (salt_count + pepper_count)
+    np.testing.assert_allclose(actual_noise_prob, prob, atol=1e-2)
+    np.testing.assert_allclose(actual_salt_to_pepper_prob, salt_to_pepper_prob, atol=1e-1)
+
+def _testimpl_operator_noise_salt_and_pepper(device, per_channel, prob, salt_to_pepper_prob, channel_first, salt_val, pepper_val, batch_size, niter):
+    pipe = pipe_salt_and_pepper_noise(prob, salt_to_pepper_prob, channel_first, per_channel,
+                                      salt_val, pepper_val,
+                                      device=device, batch_size=batch_size,
+                                      num_threads=3, device_id=0, seed=12345)
+    pipe.build()
+    salt_val = 255 if salt_val is None else salt_val
+    pepper_val = 0 if pepper_val is None else pepper_val
+    for _ in range(niter):
+        out_data, in_data, prob_arg, salt_to_pepper_prob_arg = pipe.run()
+        prob_arg = prob_arg.as_array()
+        salt_to_pepper_prob_arg = salt_to_pepper_prob_arg.as_array()
+        if device == 'gpu':
+            out_data = out_data.as_cpu()
+            in_data = in_data.as_cpu()
+        for s in range(batch_size):
+            sample_in = np.array(out_data[s])
+            sample_out = np.array(in_data[s])
+            if channel_first:  # Convert back to channel-last before verifying
+                sample_out = np.transpose(sample_out, axes=(1, 2, 0))
+                sample_in = np.transpose(sample_in, axes=(1, 2, 0))
+            prob = float(prob_arg[s])
+            salt_to_pepper_prob = float(salt_to_pepper_prob_arg[s])
+            sample_ref = salt_and_pepper_noise_ref(
+                sample_in, prob, salt_to_pepper_prob, per_channel, salt_val, pepper_val)
+            psnr_out = PSNR(sample_out, sample_in)
+            psnr_ref = PSNR(sample_ref, sample_in)
+            if dump_images:
+                import cv2
+                suffix_str = f"{prob}_{salt_to_pepper_prob}_s{s}"
+                if not per_channel:
+                    suffix_str = suffix_str + "_monochrome"
+                cv2.imwrite(f"./snp_noise_ref_p{suffix_str}.png", cv2.cvtColor(sample_ref, cv2.COLOR_BGR2RGB))
+                cv2.imwrite(f"./snp_noise_out_p{suffix_str}.png", cv2.cvtColor(sample_out, cv2.COLOR_BGR2RGB))
+            verify_salt_and_pepper(sample_out, sample_in, prob, salt_to_pepper_prob, per_channel, salt_val, pepper_val)
+            np.testing.assert_allclose(psnr_out, psnr_ref, atol=1)
```

Comment:
I'm not sure if we need that check if you already has verify_salt_and_pepper which check the result more directly.

Reply:
It has value, for several reasons:
1. I can produce a reference and verify the results make sense visually.
2. The verify_salt_and_pepper is not 100% reliable, imagine images with a lot of black or white pixels.